realm/realm-library/src/androidTest/kotlin/io/realm/RealmModelManagedSetTester.kt

```diff
@@ -0,0 +1,397 @@
+            }
+
+            // Checks it does not contain nulls
+            assertFalse(set.containsAll(listOf(null)))
+        }
+
+        // Test with objects from another realm
+        looperThread.runBlocking {
+            Realm.getInstance(realm.configuration).use { alternativeRealm ->
+                alternativeRealm.executeTransaction {
+                    val values = alternativeRealm.copyToRealmOrUpdate(unmanagedInitializedSet)
+
+                    assertFailsWith<IllegalArgumentException>("Cannot pass an object from another Realm instance") {
+                        set.containsAll(values)
+                    }
+                }
+            }
+            looperThread.testComplete()
+        }
+    }
+
+    override fun addAll() {
+        // Test with managed objects
+        managedTester.addAll()
+
+        val set = initAndAssertEmptySet()
+        realm.executeTransaction { transactionRealm ->
+            // Changes after adding collection
+            assertFailsWith<IllegalArgumentException>("Cannot add null values into a set") {
+                set.addAll(listOf(null))
+            }
+
+            assertTrue(set.addAll(unmanagedInitializedSet))
+            assertEquals(unmanagedInitializedSet.size, set.size)
+
+            // Does not change if we add the same data
+            assertFalse(set.addAll(unmanagedInitializedSet))
+            assertEquals(unmanagedInitializedSet.size, set.size)
+
+            // Does not change if we add itself to it
+            assertFalse(set.addAll(set))
+            assertEquals(unmanagedInitializedSet.size, set.size)
+
+            // Does not change if we add an empty collection
+            assertFalse(set.addAll(listOf()))
+            assertEquals(unmanagedInitializedSet.size, set.size)
+
+            // Throws when adding a collection of a different type
+            val somethingEntirelyDifferent = unmanagedInitializedSet.map {
+                Pair(it, it)
+            }
+            assertFailsWith<ClassCastException> {
+                set.addAll(somethingEntirelyDifferent as Collection<T>)
+            }
+
+            // Does not change if we add the same data from a managed set
+            val sameValuesManagedSet = managedSetGetter.get(transactionRealm.createObject())
+            assertNotNull(sameValuesManagedSet)
+            assertTrue(sameValuesManagedSet.addAll(unmanagedInitializedSet))
+            assertFalse(set.addAll(sameValuesManagedSet as Collection<T>))
+            assertEquals(unmanagedInitializedSet.size, set.size)
+
+            // Does not change if we add an empty RealmSet
+            val emptyManagedSet = managedSetGetter.get(transactionRealm.createObject())
+            assertTrue(emptyManagedSet.isEmpty())
+            assertFalse(set.addAll(emptyManagedSet))
+            assertEquals(unmanagedInitializedSet.size, set.size)
+
+            // Changes after adding a managed set containing other values
```

Comment:
I think we would also need a test showing that adding the same unmanaged objects without primary keys twice actually adds multiple copies, as core uses the object id and wouldn't consider them the same.